File: backend/userservice/endpoint/all-profiles.go
package UserEndpoint

import (
	"context"
	"encoding/json"
	"log"
	"net/http"
	"services/userservice"
	"time"
)

// handler for retrieving all profiles from firebase. Returns handler function.
func GetAllProfilesHandler(svc *userservice.Service) http.HandlerFunc {
	//actual handler function
	return func(w http.ResponseWriter, r *http.Request) {
		//has to be get method
		if r.Method != http.MethodGet {
			http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
			return
		}

		//set timeout to get context
		ctx, cancel := context.WithTimeout(r.Context(), 15*time.Second)
		defer cancel()

		users, err := svc.GetAllUserProfiles(ctx)
		if err != nil {
			log.Printf("error fetching all users: %v", err)
			http.Error(w, "error fetching all users", http.StatusInternalServerError)
			return
		}

		w.Header().Set("Content-Type", "application/json")
		if err := json.NewEncoder(w).Encode(users); err != nil {
			log.Printf("Error encoding response: %v", err)
			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
			return
		}
	}
}

File: web/pages/api/user/all-profiles.ts
import { NextApiRequest, NextApiResponse } from "next";
import { getAllUserProfiles } from "@/lib/firebase/adminUtils/userService";

//fetch all profiles from firebase.
export default async function handler(req: NextApiRequest, res: NextApiResponse) {

    if(req.method !== 'GET'){
        return res.status(405).json({ message: "Method Not Allowed." });
    }

    try{
        const userProfiles = await getAllUserProfiles();
        return res.status(200).json(userProfiles);

    } catch (err) {
        console.error("An error occurred while fetching all user profile data", err);
        return res.status(500).json({ error: "Failed to fetch all user profile data" });
    }
}